File: ts/src/emitter/mermaid/render-top-level-nodes.ts
import { VISUAL_ELEMENT_TYPE } from "../../visual-graph/visual-element-type.js";
import type { VisualGraph } from "../../visual-graph/visual-graph.js";
import { emitNode } from "./emit-node.js";
import type { RenderState } from "./render-state.js";
import { rendersInSyntheticBlock } from "./renders-in-synthetic-block.js";

export function renderTopLevelNodes(
  state: RenderState,
  graph: VisualGraph,
): void {
  for (const e of graph.elements) {
    if (
      e.type === VISUAL_ELEMENT_TYPE.Node &&
      !rendersInSyntheticBlock(e) &&
      !state.wrappedOwnerIds.has(e.id)
    ) {
      emitNode(state, e, "  ");
    }
  }
}

File: crates/unsnarl-emitter-mermaid/src/render_top_level_nodes.rs
//! Emits top-level "tree" nodes (anything that isn't a synthetic
//! top-level import/module/sink and isn't a wrapped function-owner
//! name).
//!
//! Mirrors `ts/src/emitter/mermaid/render-top-level-nodes.ts`.

use unsnarl_visual_graph::visual_element::VisualElement;
use unsnarl_visual_graph::visual_graph::VisualGraph;

use crate::emit_node::emit_node;
use crate::render_state::RenderState;
use crate::renders_in_synthetic_block::renders_in_synthetic_block;

pub fn render_top_level_nodes(state: &mut RenderState<'_>, graph: &VisualGraph) {
    for e in &graph.elements {
        if let VisualElement::Node(n) = e {
            if !renders_in_synthetic_block(n) && !state.wrapped_owner_ids.contains(n.id()) {
                emit_node(state, n, "  ");
            }
        }
    }
}
